Maintain more structural information in type-checking errors

antalsz opened the pull request:
### Introduction

This PR refactors the types that make up type-checking errors to maintain more structural information.  Most significantly, we split out the errors we get from unification, moregen, and type equality in `Ctype` into three distinct exceptions, and retain the sorts of information we keep during unification in the other two cases as well (no more `Unify []`).  We also maintain more information in `Includecore` about mismatches of primitives, manifests, and private objects and variants

This PR doesn't make a lot of changes to the user-visible error messages yet, only what's necessary to support the type changes, but it enables a forthcoming patch where we use this improvement to produce more informative error messages.

This PR builds heavily on work done by Mekhrubon Tuarev (@MekhrubonT) during his internship at Jane Street in 2019; most of the tricky work of figuring out where to insert these new exception calls was his, and I built on top of this.  Both of us were assisted by @lpw25.

### Motivation

In the past, the OCaml compiler used `Unify` as its major exception in most cases; it would `raise (Unify [])` from places like moregen and type equality where unification wasn't what was being called.  This both lost information and also didn't let us distinguish between (1) errors intentionally raised by moregen/type equality and (2) errors raised by calls to `Unify` as *implementation details* of moregen/type equality, which accidentally leaked out.  It also led to the use of `Unify []` as a general signifier for "this type-checking–related function had a problem of some sort", which meant it was possible for code to accidentally handle more errors than it expected to by assuming that the wrong function was raising the `Unify`.

### Eventual goals

The plan is to eventually use this patch to improve the actual text of errors during various kinds of module inclusion.  Right now, those errors are "shallow"; we would like to make them "deep".  For example, consider the following example in current OCaml:

    module M : sig
      val x : bool * int
    end = struct
      let x = false , "not an int"
    end

This produces the following error:

    Error: Signature mismatch:
           Modules do not match:
             sig val x : bool * string end
           is not included in
             sig val x : bool * int end
           Values do not match:
             val x : bool * string
           is not included in
             val x : bool * int

This error omits a key detail, though: the reason the signatures don't match is that `string` is not `int`.  And we *would* get that error if we used an ordinary type annotation:


    let y = false, "still not an int" in
    (y : bool * int)

This gives the more useful error

    Error: This expression has type bool * string
           but an expression was expected of type bool * int
           Type string is not compatible with type int 

In particular, note the last line.  The eventual goal of this patch is to add "Type string is not compatible with type int" (and similar messages) to the module error above (and similar errors).

Again, this patch doesn't make that change, it just lays the groundwork, with one exception: the following test case from `testsuite/tests/typing-poly/poly.ml`, lines 970–980.  The code is

    type ('a,'b) t constraint 'a = 'b
                   constraint 'a = int
      and 'a u = (float,string) t;;


This now produces the following error:

    Line 3, characters 13-29:
    3 |   and 'a u = (float,string) t;;
                     ^^^^^^^^^^^^^^^^
    Error: Constraints are not satisfied in this type.
           Type (float, string) t should be an instance of (int, int) t
           Type float is not compatible with type int

The last line, "Type float is not compatible with type int", was added by this patch.

### Changes

This PR makes the following changes:

* Move `Ctype.Unification_trace` (often bound locally as `Trace`) into the new module `Errortrace`, and make it a GADT differentiating between unification traces and non-unification traces (i.e., equality or moregen traces).

* Ensure that the code in `Ctype` correctly raises these the new `Moregen` and `Equality` exceptions, building a trace much like during unification.  Note that we do still raise `Moregen []` and `Equality []` sometimes, since our goal is for the improvement of error messages (rather than the necessary precondition of improving the machinery) to be a separate PR for easier comprehensibility.

* Be more intentional in handling `Unify` errors that arise during implementation.  This shows up in two ways: (1) hiding that unification was used as an implementation detail, and (2) not raising `Unify []` as a general "there was a problem" signifier.  For an example of (1), `Ctype.subst` uses unification internally to handle the case of a type with constraints.  However, previously, if that failed – a situation which can only happen in rare corner cases – a `Unify` exception would come out of substitution, which other functions might not expect.  Now, `subst` instead raises the new error `Cannot_subst`, which we handle wherever `subst` is called; a mistake will result in an unexpected `Cannot_subst` error, instead of an incorrectly-handled `Unify` error.

* Maintain more structural information in `Includecore`, specifically about mismatches of primitives, manifests, and private objects and variants.

* Add many new tests, most of which point at behavior that we changed but some of which point at behavior that we did not change.  We expect more of these changes in the forthcoming patch that takes advantage of the new exception structure.

The changes are structured into two commits: first, a commit that adds new tests; and second, a commit that makes the actual code changes and updates all the necessary tests.

### Detailed, file-by-file changes

This section documents the changes we made to every file, since we touch so many of them; it is not required reading, but hopefully helps review.  There is nothing after this section in this PR, so if you don't want to read the gory details, you can stop now.

The key files to look at are, in roughly descending order, `typing/ctype.ml`, `typing/printtyp.ml`, `typing/errortrace.ml`, and `typing/includecore.ml`, as well as, to a lesser extent, their `.mli` files.

* `.depend` *(modified)*:
  
  Inessential changes, autogenerated

* `compilerlibs/Makefile.compilerlibs` *(modified)*:
  
  Added the new `Errortrace` module

* `lambda/lambda.ml` *(modified)*:

  Moved `equal_boxed_integer` to `primitive.ml`.

* `testsuite/tests/typing-misc/constraints.ml` *(modified)*:

  New tests for `constraint` errors; these errors highlight old behavior that hasn't changed, not new behavior.  However, because we fix an indentation bug, the tests we added, as well as one other test, show up in the diff.

* `testsuite/tests/typing-misc/pr6416.ml` *(modified)*:

  Changed the error message output here, removing some arguably-redundant text that was leaking out of `subst` (see the discussion in `Ctype.ml`).

* `testsuite/tests/typing-misc/pr7937.ml` *(modified)*:

  Deleted a bunch of spurious "Types for tag \`X are incompatible" lines from errors; they were showing up even when only one side was a variant, and now they're gone instead

* `testsuite/tests/typing-misc/unbound_type_variables.ml` *(new)*:

  New tests about unbound type variables; these errors highlight old behavior that hasn't changed, not new behavior.  One error has an indentation fix.

* `testsuite/tests/typing-modules/inclusion_errors.ml` *(new)*:

   New tests that confirm our changes to moregen and type equality (in `Ctype`), and our changes to manifest checking (in `Includecore`) do not change existing behavior.  The file has three sections:
   1. First, it checks that our new `Equality` exception is handled like `Unification` was previously.
   2. Second, it similarly checks that our new `Moregen` exception is handled like `Unification` was previously.
   3. Finally, tests for the new changes in `Includecore` that provide improved handling of the situation when manifests don't match (a manifest, also known as an equation, is the `s` in `type t = s = X | Y` or `type t = s`).

* `testsuite/tests/typing-objects-bugs/pr3968_bad.compilers.reference` *(modified)*:

  In making "is more general than" errors consistent with "unify" errors, this test output has one more line at the end saying "Types for tag ``App` are incompatible".  This refers to the tag *two* errors before the message, not *one* more.  Changes to improve this error could (should?) be made, but are deferred to making changes that improve unify errors more generally.

* `testsuite/tests/typing-objects-bugs/pr4018_bad.compilers.reference` *(modified)*:

  Reduced a lot of noise in this error message that was due to a `Unify` error escaping when a substitution failed due to constraints (`ctype.ml:1456-1459`, in `Ctype.subst`).

* `testsuite/tests/typing-poly/poly.ml` *(modified)*:

  Add one new test that confirms that we're maintaining more of an error trace, and reword one test to have clearer names

* `testsuite/tests/typing-poly/pr9603.ml` *(modified)*:

  There is a known bug with polymorphic methods where unifying with universal variables is over-aggressively forbidden.  This error message was cut short due to `Ctype.subst` no longer revealing that it uses unification internally; the previous text was not particularly helpful, but the new error text is not particularly more helpful, just shorter.

* `testsuite/tests/typing-polyvariants-bugs-2/pr3918c.compilers.reference` *(modified)*:

  Reduce the text in the error message that occurs when we cannot find interface files.  The error message didn't make sense before ("the type variable `'a` occurs inside [an expression without `'a`]") and still doesn't make sense now ("This expression has type `'b Pr3918b.vlist` but an expression was expected of type `'b Pr3918b.vlist`" – the same thing twice).

* `toplevel/byte/topdirs.ml` *(modified)*:

  Changed `Ctype.equal` to `Ctype.is_equal`; see `typing/ctype.ml` for details.

* `toplevel/genprintval.ml` *(modified)*:

  Changed `Ctype.moregeneral` to `Ctype.is_moregeneral`; see `typing/ctype.ml` for details.

* `typing/ctype.ml` *(modified)*:

  As in `ctype.mli`, we remove `Unification_trace`, moving it to `Errortrace` and improving it, and add several new exceptions.  `Unification_trace` was also bound locally as `Trace`; we do not preserve the synonym.  We also add some local exceptions, which we use to ensure that our handlers are not overbroad, and some local functions for creating the new `Escape` exception.
   
  We make some common changes, such as:
  + Bringing in `Errortrace`.
  + Changing which exceptions we raise and catch to use the new exception types: `Trace.scope_escape` becomes `scope_escape`, `Trace.(Unify [escape ...])` becomes `escape_exn ...`, etc.
  + Relatedly, we define a new `trace_exn` type, which allows us to specify which of the three trace exceptions (`Unify`, `Moregen`, or `Equality`) we should raise at runtime.  We switch all raises of `Unify` to use the associated `raise{_nil,1}_for` functions and, when we change `moregen` and `equal` to raise their new associated exceptions, use `raise{_nil,1}_for` as well.  The real win from defining this type, however, is that when we define wrappers around utility functions to rewrap their errors from bespoke errors to trace errors, we need define only one wrapper, rather than two or three (for instance, we define `occur_for` which wraps `occur`, and then can use both `occur_for Unify` and `occur_for Moregen`).
  + Replacing `try_expand_once` with `try_expand_safe` nearly everywhere, since the latter throws fewer exceptions.
  + We change the definition of `expand_head_opt` to be more consistent with the behavior of `expand_head`.
  + Replacing various occurrences of `unify_univar` with `link_univar`, since it throws a more specific exception (see below).
   
  In `subst`, we correctly forget type abbreviations, which fixes a bug.  Before, abbreviations which were tentatively expanded during a failed substitution would stick around.  Substitution can only fail like this because of strange type constraints.  We also swallow the unification trace during these failures, as the fact that substitution failed because of unification is an implementation detail; this shortens unhelpful error messages in a few places.
   
  The local function `unify_univar` becomes `link_univar`, and raises the new `Cannot_unify_universal_variables` exception instead of a generic `Unify` exception; `unify_univar` is still provided, raising the old exception.
   
  We define `equal_private`, a variant of `equal` that on failure tries again after looking through private types.
   
  We replace `equal_clty` with `equal_clsig`, which contains the only portion of the old `equal_clty` that we actually used.

* `typing/ctype.mli` *(modified)*:

  Remove `Unification_trace`, moving it to `Errortrace` and improving it, and add several new exceptions: `Equality` and `Moregen`, which behave like `Unification`; `Escape`; and `Matches_failure`.  We also improve the `Subtype` exception.
   
  On the value side, we continue to manage exceptions: first, we replace `try_expand_once` with the safer `try_expand_safe`, which can never raise `Escape` (formerly `Unify`).  More generally, we change `moregeneral`, `matches`, and `equal` to return `unit`, *raising an exception* on failure.  The old behavior is available through `is_moregeneral` and `is_equal` (we do not provide `does_match` as it wasn't necessary).
   
  Finally, we add `equal_private`, a type equality that can look inside private types; this is used in our changes to `Includecore`.

* `typing/errortrace.ml` *(new)*:

  A new module, replacing `Ctype.Unification_trace` (sometimes bound locally as `Trace`).  See `typing/ctype.ml` for more details.
   
  In our refactoring, we discovered that unification and moregen can both throw `Rec_occur`, and unification, moregen, and equality can all three throw `Escape`.  We don't quite see how moregen and equality can throw these exceptions – we expect these type-mutation–related errors to only show up during proper unification.  But we're not absolutely convinced those code paths will never fire.  Either a patch to remove those constructors demonstrating that this is safe, or a test case proving that they're necessary, would be a good item of future work.

* `typing/errortrace.mli` *(new)*:

  A new module, replacing `Ctype.Unification_trace` (sometimes bound locally as `Trace`).  See `typing/ctype.ml` for more details.

* `typing/includeclass.ml` *(modified)*:

  Two changes: (1) switch from `Printtyp.report_unification` to `Printtyp.report_error` with the appropriate trace format; (2) Add basic printing for the new equality-error cases `CM_Val_type_mismatch_eq` and `CM_Meth_type_mismatch_eq`.

* `typing/includecore.ml` *(modified)*:

  Improve the handling of mismatches of primitives, manifests, and private objects and variants.  This involves incorporating values of appropriate `Errortrace.t` types, as well as maintaining new structural information in this file.  For example, we now maintain the reason two primitives are not the same in a `primitive_mismatch` type that records the difference.  The function `type_manifest` is also moved much later in the file and simplified by having much of its contents factored out.

* `typing/includecore.mli` *(modified)*:

  Maintain extra structural information about mismatches of primitives, manifests, and private types.

* `typing/includemod.ml` *(modified)*:

  Save the information from the new `Includecore` errors, but don't print it yet.  We pack an `Includecore.value_mismatch` value into `Value_descriptions`, but then ignore it when pattern-matching.

* `typing/includemod.mli` *(modified)*:

  Save the information from the new `Includecore` errors in the `Value_descriptions` case.

* `typing/primitive.ml` *(modified)*:

  Equality functions used in `Includecore`; `equal_boxed_integer` comes from `lambda.ml`.

* `typing/primitive.mli` *(modified)*:

  Equality functions used in `Includecore`; `equal_boxed_integer` comes from `lambda.ml`.

* `typing/printtyp.ml` *(modified)*:

  Here is where we apply the printing change.  This is actually smaller than might otherwise be expected; the changes to printing are largely around making everything *more uniform*.  That said, there is certainly churn here, including noise from replacing `Trace` (i.e., `Ctype.Unification_trace`) with `Errortrace` everywhere.
   
  One big change is that we introduce the type `trace_format`, which configures trace printing, and provide one value of this type for each kind of trace: `unification : unification trace_format`, `equality : non_unification trace_format`, and `moregen : non_unification trace_format`.  In the end, the user will call `report_error unification`, `report_error equality`, or `report_error moregen` as necessary.
   
  Various functions are generalized by taking this extra parameter, in order to configure their behavior.  Other functions are generalized to support ordinary traces or subtyping traces (e.g. `prepare_any_trace`, which can handle both kinds, vs. `prepare_trace`, which is specialized to ordinary traces).
   
  The `print_pos` function moved to `Errortrace`, to be with the definition of the type.
   
  Most of the explanation functions retained their old behavior, but there is visual churn in the diff due to changing module names and indentation (the latter hopefully minimized, but occasionally unavoidable).
   
  The new outwards-facing entry point is principally `report_error`, which is designed to support all three major traces; similarly, we have `Subtyping.report_error` for subtyping traces.  The `Subtyping` module in general now holds all subtyping-trace–related code in order to use the same names and be self-contained.  Someday, I hope that we can unify it more fully.

* `typing/printtyp.mli` *(modified)*:

  Replace `report_unification_error` and `report_subtyping_error` with `report_error` and `Subtype.report_error`, respectively; the former is parameterized by a new `trace_format` type that controls which sort of trace (unification, equality, or moregen) to use.

* `typing/typeclass.ml` *(modified)*:

  Replace `Ctype.Unification_trace.t` with `Errortrace.unification Errortrace.t`, drop reference to `Ctype.Unification_trace`, replace `Printtyp.report` with `Printtyp.report_error`.  One error message got more of a printing box placed around it.

* `typing/typeclass.mli` *(modified)*:

  Replace `Ctype.Unification_trace.t` with `Errortrace.unification Errortrace.t` everywhere.

* `typing/typecore.ml` *(modified)*:

  Upgrade to the new errors: replace `Ctype.Unification_trace` with `Errortype`, replace `Printtyp.report_unification_error` with `Printtyp.report_error`, etc.

* `typing/typecore.mli` *(modified)*:

  Upgrade to the new errors: replace `Ctype.Unification_trace` with `Errortype`.

* `typing/typedecl.ml` *(modified)*:

  Upgrade to the new errors.  The new `Env.t` in the `Constraint_failed` case enables us to lose less information.  We also generally improve printing (in addition to the use of the new `Printtyp.report_error`): we fix the indentation of one error message by changing `@.` to `@ `, improve the printing in the `Constraint_failed` case (where we now have more information), and use more formatting boxen.

* `typing/typedecl.mli` *(modified)*:

  Upgrade to the new errors: replace `Ctype.Unification_trace.t` with `Errortype.t`.  Also, in the `Constraint_failed` case, replace a pair of `type_expr`s with an `Env.t` and an `Errortrace.t`.

* `typing/typedecl_variance.ml` *(modified)*:

  Change `Ctype.equal` to `Ctype.is_equal`; see `ctype.ml` for more information.

* `typing/typetexp.ml` *(modified)*:

  Replace `Ctype.Unification_trace` with `Errortrace`, `Ctype.equal` with `Ctype.is_equal`, and `Printtyp.report_unification_error` with `Printtyp.report_error`.

* `typing/typetexp.mli` *(modified)*:

  Replace `Ctype.Unification_trace.t` with the appropriate `Errortrace.t`

Commit:
Add tests before improving the internal type error representation

Some of these tests point at behavior we're about to change; some of
them point to behavior we're confirming that we don't change.

* `constraints.ml`, `unbound_type_variables.ml`: Existing code whose
  behavior we won't change; however, we do fix the indentation in some
  of the type errors.
* `poly.ml`: Add one test whose behavior will change (more detail) and
  reword one test to use clearer variable names.
* `inclusion_errors.ml`: New tests that will confirm our internal
  changes did not (yet) have externally-visible changes to the type
  errors.

This commit includes existing work by Mekhrubon Tuarev, who did all
the early development.  I (Antal Spector-Zabusky) came and built the
layer of types on top.

antalsz committed:
Maintain more structural information in type-checking errors

This commit refactors the types that make up type-checking errors to
maintain more structural information.  Most significantly, we split
out the errors we get from unification, moregen, and type equality in
`Ctype` into three distinct exceptions, and retain the sorts of
information we keep during unification in the other two cases as well
(no more `Unify []`).  We also maintain more information in
`Includecore` about mismatches of primitives, manifests, and private
objects and variants

This commit doesn't make a lot of changes to the user-visible error
messages yet, only what's necessary to support the type changes, but
it enables a forthcoming patch where we use this improvement to
produce more informative error messages.

This commit includes existing work by Mekhrubon Tuarev, who did all
the early development.  I (Antal Spector-Zabusky) came and built the
layer of types on top.

Octachron commented on typing/printtyp.ml:
```diff
@@ -1959,6 +1995,8 @@ let may_prepare_expansion compact (t, t') =
       mark_loops t; (t, t)
   | _ -> prepare_expansion (t, t')
 
+let print_path f p = !Oprint.out_ident f (tree_of_path Type p)
```
Note: this function is much more difficult to use correctly that it appears: the construction of `tree_of_path` is effectful and mutate the global state of `short-path` and the naming context. It is quite often not the right thing to delay the construction of the outcome tree to the printing time.

An easier function to use correctly would be
```ocaml
let print_path p = Format.dprintf "%a" !Oprint.out_ident (tree_of_path Type p)
```
This variant build the tree immediately (in the current context) and only delay its printing.

lpw25 commented on typing/errortrace.ml:
```diff
@@ -0,0 +1,139 @@
+  Incompatible_fields { name; diff={got; expected} }
+
+
+let swap_unification_elt = function
```
Might as well just have `swap_trace` since swapping an equality trace makes sense. Should also avoid the fragile match.

lpw25 commented on typing/includeclass.ml:
```diff
@@ -70,23 +70,35 @@ let include_err ppf =
           fprintf ppf "The instance variable %s@ has type" lab)
         (function ppf ->
           fprintf ppf "but is expected to have type")
+  | CM_Val_type_mismatch_eq (lab, env, trace) ->
```
I think that we can probably get away without a new constructor for this now: just a boolean or two case variant type saying whether the error was from `moregen` or `eqtype` will suffice.

lpw25 commented on typing/errortrace.mli:
```diff
@@ -0,0 +1,97 @@
+
+  val diff: type_expr -> type_expr -> desc elt
+
+  val map : (desc -> desc) -> desc elt list -> desc elt list
```
Maybe put after flatten to match the order in the outer module.

lpw25 commented on typing/printtyp.ml:
```diff
@@ -1903,36 +1900,75 @@ type printing_status =
-  | [] -> []
-  | elt :: rem -> (* the first element is always kept *)
-      elt :: List.fold_right clean_trace rem []
+  let rec prepare_trace_helper tr = match tr with
```
Putting the match on the same line like this is an unusual style, which I think should be avoided, even though it is used in the existing code a few lines above. In this case it is a particularly odd style since you could just use `function` instead of `match`.

The name `prepare_trace_helper` is both meaninglessly generic and long, it should probably only be one or the other. I personally like `loop` but anything like that is fine.

lpw25 commented on typing/printtyp.ml:
```diff
@@ -1995,118 +2033,120 @@ let explanation_diff env t3 t4 : (Format.formatter -> unit) option =
+  | Errortrace.Self ->
       Some (dprintf "%t@,Self type cannot escape its class" pre)
+  | Errortrace.Constraint ->
+      Some (dprintf "%t@,###########Constraint##########" pre)
```
This looks like some left over debug code. I don't see anything in the test suite that is producing this message, so either there is a missing test case or this should be `assert false`.

Octachron replied:
All the corresponding cases in the testsuite are currently suppressed by the `constraint_escape_status` setting.

antalsz replied:
I've removed `constraint_escape_status` and replaced this case with `None`.

lpw25 commented on typing/printtyp.ml:
```diff
@@ -1995,118 +2033,120 @@ let explanation_diff env t3 t4 : (Format.formatter -> unit) option =
+  | Errortrace.Diff { Errortrace.got = _,s; expected = _,t } ->
+    explanation_diff env s t
+  | Errortrace.Escape {kind;context} ->
+    let pre = match context, kind, prev with
```
Should have a line break after `=`

lpw25 commented on typing/printtyp.ml:
```diff
@@ -1995,118 +2033,120 @@ let explanation_diff env t3 t4 : (Format.formatter -> unit) option =
+                 but the expected method type was@ %a@]"
+          name type_expr diff.got type_expr diff.expected
+      | _ -> ignore
+    in explain_escape pre kind
```
Should have a line break after `in`

lpw25 commented on typing/printtyp.ml:
```diff
@@ -1995,118 +2033,120 @@ let explanation_diff env t3 t4 : (Format.formatter -> unit) option =
+  | Errortrace.Obj o ->
+    explain_object o
+  | Errortrace.Rec_occur(x,y) ->
+    mark_loops y;
```
This seems to have changed from `reset_and_mark_loops`.

lpw25 commented on typing/printtyp.ml:
```diff
@@ -1995,118 +2033,120 @@ let explanation_diff env t3 t4 : (Format.formatter -> unit) option =
+      | Some ctx, _, _ ->
+        dprintf "@[%t@;<1 2>%a@]" intro type_expr ctx
+      | None, Univ _, Some(Errortrace.Incompatible_fields {name; diff})
+        when trace_format.explain_contextless_escaped_field_mismatch ->
```
I don't think we actually need this part to be different between the different sorts of traces. I think that the difference probably arose due to the original patch being rebased across the PR that added this code rather than through a deliberate choice.

Octachron replied:
This seems even more true since the `Univ _` is not expected to happen in `Moregen` and `Equality`.

lpw25 commented on typing/printtyp.ml:
```diff
@@ -1903,36 +1900,75 @@ type printing_status =
+
+let printing_status trace_format = function
+  | Errortrace.Diff d -> diff_printing_status d
+  | Errortrace.Escape {kind = Constraint} -> trace_format.constraint_escape_status
```
I don't think this difference between the sorts of traces is needed. I think we can just always `Keep` the constraint escapes, as long as we also fix the message for them.

lpw25 commented on typing/printtyp.ml:
```diff
@@ -1903,36 +1900,75 @@ type printing_status =
-      elt :: List.fold_right clean_trace rem []
+  let rec prepare_trace_helper tr = match tr with
+    | [] -> []
+    | elt :: rem when drop elt = true -> prepare_trace_helper rem
```
I'm not sure what this `drop` function is for. I think possibly we can just get rid of `drop_from_trace`.

lpw25 commented on typing/printtyp.ml:
```diff
@@ -1903,36 +1900,75 @@ type printing_status =
+  else if same_path t1 t1' && same_path t2 t2' then Optional_refinement
+  else Keep
+
+type 'variety trace_format = {
```
I think that my other review comments mean that this type only needs the `incompatibility_phrase` field. I think that having a `trace_format` type is still a good idea, since we may well want to adjust further printing behaviour based on the sort of trace, but it might be worth just switching it to a simple three case GADT and making `incompatiblity_phrase` a function out of that type rather than a record field.

Octachron replied:
I agree with this proposed simplification.

antalsz replied:
I went with an approach where it is implemented as a single-field record and left abstract; this can then easily be changed in the future.

antalsz replied:
I updated this again to used the proposed approach, which dose seem nicer.  I also no longer expose this type, because we instead expose three separate `report_{unify,equality,moregen}_error` functions.  This can still be changed in the future if things become more appealing to do that way.

lpw25 commented on typing/printtyp.ml:
```diff
@@ -2176,51 +2218,96 @@ let unification_error env tr txt1 ppf txt2 ty_expect_explanation =
-          | _ -> [] in
+
+module Subtype = struct
+  (* There's a frustrating amount of code duplication between this module and
```
This comment is out-of-date: there is no [Make_trace_printer].

lpw25 commented on typing/errortrace.ml:
```diff
@@ -0,0 +1,139 @@
+  | Escape { kind = Equation x; context} -> Escape { kind = Equation (f x); context }
+  | Escape { kind = (Univ _ | Self | Constructor _ | Module_type _ | Constraint); _ }
+  | Variant _ | Obj _ | Incompatible_fields _ as x -> x
+  | Rec_occur (_, _) as x -> x
```
No need for the `Rec_occur` case to be separate from the one above it.

lpw25 commented on typing/printtyp.ml:
```diff
@@ -2176,51 +2218,96 @@ let unification_error env tr txt1 ppf txt2 ty_expect_explanation =
-    print_labels := true;
-    raise exn
+        if fst then trace fst txt ppf (elt @ tr)
+        else trace fst txt ppf tr;
```
I think it would clearer to rename `map_diff` to `get_diff` and have it return an option, and then do something like:
```ocaml
        let tr =
          trees_of_trace
            (List.map (Errortrace.map_diff prepare_expansion)
               (filter_trace keep_last tr'))
        in
        let tr =
          match fst, get_diff elt with
          | true, Some elt -> elt :: tr
          | _, _ -> tr
        in
        trace fst txt ppf tr
```

antalsz replied:
A note for posterity: the ordered mattered here.  Moving `get_diff` after the `trees_of_trace ...` call broke a test case.

lpw25 commented on typing/printtyp.ml:
```diff
@@ -2126,38 +2166,40 @@ let warn_on_missing_def env ppf t =
-  reset ();
-  let tr = prepare_trace (fun t t' -> t, hide_variant_name t') tr in
-  let mis = mismatch txt1 env tr in
+let handle_trace
```
This function has only a single call, which is in tail position. I think it would be clearer to just inline it there.

lpw25 commented on typing/printtyp.ml:
```diff
@@ -2176,51 +2218,96 @@ let unification_error env tr txt1 ppf txt2 ty_expect_explanation =
-  wrap_printing_env env (fun () -> unification_error env tr txt1 ppf txt2
-                            type_expected_explanation)
+let error trace_format env tr txt1 ppf txt2 ty_expect_explanation =
+  let prepare_expansion_head empty_tr = function
```
I think this function can be lifted back to where it was before.

antalsz replied:
I couldn't figure out where this was supposed to be?

antalsz replied:
I was looking at `error`, instead of `prepare_expansion_head`.  This is done now.

lpw25 commented on typing/printtyp.ml:
```diff
@@ -2176,51 +2218,96 @@ let unification_error env tr txt1 ppf txt2 ty_expect_explanation =
-  )
+  let filter_unification_trace = filter_trace unification
+
+  let rec filter_trace keep_last = function
```
Probably better to call this `filter_subtype_trace`

lpw25 commented on typing/ctype.ml:
```diff
@@ -56,109 +57,41 @@ open Local_store
+| Moregen  : non_unification trace_exn
+| Equality : non_unification trace_exn
+
+let raise_for (type variant) (tr_exn : variant trace_exn) (tr : variant Errortrace.t) : 'a =
```
Since it's the least commonly used, maybe call this `raise_trace_for` and rename `raise1_for` to `raise_for`.

lpw25 commented on typing/ctype.ml:
```diff
@@ -2263,7 +2244,7 @@ let rec mcomp type_pairs env t1 t2 =
             begin try
               let decl = Env.find_type p env in
-              if non_aliasable p decl || is_datatype decl then raise (Unify [])
+              if non_aliasable p decl || is_datatype decl then raise_nil_for Unify
```
Since it doesn't actually keep a trace, I think it would be better for `mcomp` to raise its own `Incompatible` exception, and then have an `mcomp_for` that raises an error. Don't forget to change the uses of `link_univar` and `enter_poly_for` as well.

lpw25 commented on typing/ctype.ml:
```diff
@@ -56,109 +57,41 @@ open Local_store
+(* Uses of this function are a bit suspicious, as we usually want to maintain trace
+   information; sometimes it makes sense, however, since we're maintaining the trace at an
+   outer exception handler. *)
+let raise_nil_for tr_exn =
```
I suggest a more semantic name like `raise_unexplained_for`.

lpw25 commented on typing/ctype.ml:
```diff
@@ -56,109 +57,41 @@ open Local_store
+
+let escape kind = {kind; context = None}
+let escape_exn kind = Escape (escape kind)
+let scope_escape_exn kind = escape_exn (Equation (short kind))
```
These are only ever used with `raise`. Might as well replace them with `raise_scope_escape` etc.

lpw25 commented on typing/ctype.ml:
```diff
@@ -2538,6 +2519,18 @@ let unify_package env unify_list lv1 p1 n1 tl1 lv2 p2 n2 tl2 =
 (* force unification in Reither when one side has a non-conjunctive type *)
 let rigid_variants = ref false
 
+(* Only used in unify; could become [update_level_for tr_exn] if it were used elsewhere *)
```
It's also used in `moregen`, so it should indeed be `update_level_for`.

lpw25 commented on typing/ctype.ml:
```diff
@@ -2538,6 +2519,18 @@ let unify_package env unify_list lv1 p1 n1 tl1 lv2 p2 n2 tl2 =
+    update_level env level ty
+  with Escape e -> raise1_for Unify (Escape e)
+
+(* Only used in unify; could become [occur_univar_for tr_exn] if it were used elsewhere *)
```
It's also used in `moregen`, so it should indeed be `occur_univar_for`.

lpw25 commented on typing/ctype.ml:
```diff
@@ -1925,11 +1885,17 @@ let rec unify_univar t1 t2 = function
+
+(* The same as [link_univar], but raises [Unify] instead of
+   [Cannot_unify_universal_variables] *)
+let unify_univar t1 t2 univar_pairs =
```
This should become `link_univar_for` and replace all the uses of `link_univar` other than the one in `mcomp`.

lpw25 commented on typing/ctype.ml:
```diff
@@ -175,12 +108,16 @@ let () =
 
+exception Cannot_subst
+
+exception Cannot_unify_universal_variables of type_expr * type_expr
```
I don't think the arguments are actually used, so they should probably go.

lpw25 commented on typing/ctype.ml:
```diff
@@ -2641,21 +2634,24 @@ let rec unify (env:Env.t ref) t1 t2 =
-  ignore (expand_head_unif !env t2);
-  let t1' = expand_head_unif !env t1 in
-  let t2' = expand_head_unif !env t2 in
+  let t1', t2' =
```
This change is unnecessary and looks like the remnants of a change that was backed out.

lpw25 commented on typing/ctype.ml:
```diff
@@ -2807,30 +2805,30 @@ and unify3 env t1 t1' t2 t2' =
-      | (Tnil,  Tconstr _ ) -> raise (Unify Trace.[Obj(Abstract_row Second)])
-      | (Tconstr _,  Tnil ) -> raise (Unify Trace.[Obj(Abstract_row First)])
-      | (_, _) -> raise (Unify [])
+      | (Tnil,  Tconstr _ ) -> raise (Unify Errortrace.[Obj(Abstract_row Second)])
```
Maybe use `raise_for Unify`?

lpw25 commented on typing/ctype.ml:
```diff
@@ -2672,7 +2668,7 @@ and unify2 env t1 t2 =
   else
     try unify3 env t2 t2' t1 t1' with Unify trace ->
-      raise (Unify (Trace.swap trace))
+      raise (Unify (swap_unification_trace trace))
```
Maybe use `raise_trace_for Unify`?

lpw25 commented on typing/ctype.ml:
```diff
@@ -3018,9 +3016,9 @@ and unify_row_field env fixed1 fixed2 rm1 rm2 l f1 f2 =
     | Some fix ->
-        let tr = Trace.[ Variant (Fixed_row (pos,Cannot_add_tags [l],fix)) ] in
+        let tr = [Variant(Fixed_row(pos,Cannot_add_tags [l],fix))] in
         raise (Unify tr) in
```
Maybe use `raise_for Unify`

lpw25 commented on typing/ctype.ml:
```diff
@@ -3146,14 +3145,16 @@ let unify_var env t1 t2 =
+        let expanded_trace =
+          expand_trace env @@ Errortrace.diff t1 t2 :: trace
+        in
         raise (Unify expanded_trace)
```
Maybe use `raise_trace_for Unify`

lpw25 commented on typing/ctype.ml:
```diff
@@ -3001,7 +2999,7 @@ and unify_row env row1 row2 =
         try unify_row_field env fixed1 fixed2 rm1 rm2 l f1 f2
         with Unify trace ->
-          raise Trace.( Unify( Variant (Incompatible_types_for l) :: trace ))
+          raise ( Unify( Variant (Incompatible_types_for l) :: trace ))
```
Maybe use `raise_trace_for Unify`

lpw25 commented on typing/ctype.ml:
```diff
@@ -3274,11 +3275,13 @@ let moregen_occur env level ty =
+  try
+    occur_univar env ty;
+    update_level env level ty
+  with Escape e -> raise1_for Moregen (Escape e)
```
Once `occur_univar_for` and `update_level_for` exist, this exception handler can be removed.

lpw25 commented on typing/ctype.ml:
```diff
@@ -3523,12 +3556,16 @@ let matches env ty ty' =
-  in
-  backtrack snap;
-  ok
+  try
```
This would be clearer using `match ... with ... exception ...` instead of `try ... with ...`

lpw25 reviewed (changes requested):
This all looks pretty good. I've suggested various things to tidy up. Since I was closely involved with producing this PR, it would be good to get review from someone else too, even if only at a high-level. Perhaps one of @Octachron @gasche or @garrigue might be interested in having a look.

Octachron commented:
I am planning to review this PR in the upcoming weeks.

garrigue commented on typing/ctype.ml:
```diff
@@ -1911,7 +1871,7 @@ let local_non_recursive_abbrev env p ty =
    be done at meta-level, using bindings in univar_pairs *)
 (* TODO: use find_opt *)
-let rec unify_univar t1 t2 = function
+let rec link_univar t1 t2 = function
```
I don't like this naming: this is really a unification function, not a graph function.
If would prefer to keep the current name, which correctly matches this behaviour.
Why do you need to keep the original `unify_univar` ?

antalsz replied:
The naming was kept to support using `link_univar` in moregen and equality.  I've renamed everything to `unify_univar` and `unify_univar_for`, so we now have `unify_univar_for Unify`, `unify_univar_for Equality`, and `unify_univar_for Moregen`.  Do those seem like better names?

garrigue commented on typing/ctype.ml:
```diff
@@ -3393,55 +3407,72 @@ and moregen_row inst_nongen type_pairs env row1 row2 =
+  begin
+    match r1 with
+    | [] -> ()
+    | _ :: _ -> raise1_for Moregen (Variant (No_tags (Second, r1)))
```
Do you really need to use pattern-matching here?
`if r1 <> []` seems more compact and just as readable.

garrigue commented on typing/includecore.ml:
```diff
@@ -33,18 +79,24 @@ let value_descriptions ~loc env name
-          if p1 = p2 then Tcoerce_none else raise Dont_match
+  match Ctype.moregeneral env true vd1.val_type vd2.val_type with
+  | exception Ctype.Moregen trace -> raise (Dont_match (Type (env, trace)))
+  | () -> begin
```
Seeing this code, I wonder whether it would not be better to make `Ctype.moregeneral` return a `(unit, Errortrace.t) result`. It is a bit strange to turn a function that was intentionally not raising an exception into one that does. But I did not check all the uses.

antalsz replied:
We provide both `Ctype.moregeneral`, which raises, and `Ctype.is_moregeneral`, which returns a boolean, so that either raising or not raising can be accomplished as desired.

lpw25 replied:
I think it is good to have these functions match the behaviour of `unify`. We can decide later if we want to move all these functions over to using a result type.

garrigue commented on typing/errortrace.mli:
```diff
@@ -0,0 +1,97 @@
+
+(* Type indices *)
+type unification     = private Unification
+type non_unification = private Non_unification
```
`non_unification` -> `comparison` ?

antalsz replied:
This was a perfect choice of name!

garrigue commented on typing/errortrace.mli:
```diff
@@ -0,0 +1,97 @@
+  | Escape : 'a escape -> ('a, _) elt
+  | Incompatible_fields : { name:string; diff: type_expr diff } -> ('a, _) elt
+  (* Unification & Moregen; included in Equality for simplicity *)
+  | Rec_occur : type_expr * type_expr -> ('a, _) elt
```
Do really the distinctions between unification and comparison justify using GADTs all over?
We are not talking about safety critical code here, but just error reporting.
On the other hand, this does not look too heavy handed.

antalsz replied:
We found that it definitely simplified our thinking about the code, and as you say tried to avoid anything too gnarly.

garrigue reviewed (commented):
Overall, this looks very nice, and I think I like the approach.
I am a bit afraid of the GADTs, but it may be the price to pay for a nice factorization of the code.
Note that I read most of the code, but not that in `Printtyp`, which is too hard to reason about. I believe that you have checked carefully that the output is as expected.

Octachron commented on testsuite/tests/typing-poly/pr9603.ml:
```diff
@@ -29,8 +29,5 @@ Error: This expression has type
-       is not compatible with type < left : 'left0; right : 'right0 > pair
-       The method left has type 'a, but the expected method type was 'left
-       The universal variable 'left would escape its scope
+       Types for method m are incompatible
```
This seems like a regression. There is no information left in the remaining error message
whereas the sibylline preexisting message at least points to an issue with universal type variables.

lpw25 replied:
I basically disagree. The old error message is kind of nonsense, caused by one unification error leaking and being fused with another one. The new error message at least only says things that make sense.

Of course neither error message gives a real user a hope in hell of finding the actual issue, but that is because this is basically a bug. We could try to give a specific error message for this issue but, since @trefis is currently working on some changes that will fix it, I don't think it is worth the effort.

Octachron replied:
Fair enough. I agree that both error messages are basically undecipherable without some internal knowledge of the typechecker.

antalsz replied:
If we're all agreed, then shall we mark this as resolved?

Octachron replied:
Indeed.

Octachron commented on typing/ctype.ml:
```diff
@@ -19,6 +19,7 @@ open Misc
 open Asttypes
 open Types
 open Btype
+open Errortrace
```
Bikesheding: this global open seems unwarranted. Removing it would instance allow to replace `swap_unification_trace` with `Errortrace.swap`.

antalsz replied:
With the (unrelated) change from `swap_unification_trace` to `swap_trace`, this seemed less necessary to me?

Octachron replied:
Note that this is purely a question of personal taste, I prefer to have less global open when possible but you can totally ignore the comment above.

antalsz replied:
I will take your comment to heart for future code and leave it be for now, then, I think :-)

gasche commented:
@Octachron if would like this PR to get a decision, but I can't find the courage to review it myself. I'll trade a review from you on it against a review from me on #10219.

lpw25 commented:
@antalsz is in the process of updating the PR based on my and Jacques' reviews. Since Jacques already reviewed most of the PR I think we just need @Octachron to review the stuff in `Printyp`.

Octachron commented on typing/printtyp.ml:
```diff
@@ -1903,36 +1900,75 @@ type printing_status =
+  incompatibility_phrase = "is not equal to type";
+  drop_from_trace = (function
+    | Diff {got = ({desc = Tpoly _}, _); expected = ({desc = Tpoly _}, _)} -> true
+    | _ -> false);
```
Dropping the differences with an universal quantification seems not that useful. Consider:
```ocaml
module M: sig
  class type ['b] t = object
    method m: 'a. 'a -> 'b
  end
end = struct
  class type ['a] t = object
    method m: 'b. 'a -> 'b
  end
end
```
after dropping the `Tpoly` diffs, the error message becomes 

>```
>       Class type declarations do not match:
>         class type ['a] t = object method m : 'a -> 'b end
>       does not match
>         class type ['b] t = object method m : 'a -> 'b end
>       The method m has type 'a -> 'b but is expected to have type 'a0 -> 'b0
>       Type 'a is not compatible with type 'a0 
>```

which seems incomplete compared to
>```
>       The method m has type 'b. 'a -> 'b but is expected to have type
>         'a0. 'a0 -> 'b
>```

Moreover, removing this extra filtering would remove the need for `drop_from_trace` entirely.

Octachron commented on typing/printtyp.ml:
```diff
@@ -1995,118 +2033,120 @@ let explanation_diff env t3 t4 : (Format.formatter -> unit) option =
-  | Trace.Univ u ->  Some(
+  (* Equality & Moregen *)
+  | Errortrace.Openness pos ->
+    Some(dprintf "@,The %a variant is open and the %a is not"
```
Small nitpicking: `The %a variant type` is more in line with the other error messages and eliminate the potential confusion between `(ordinary) variant constructor` and `polymorphic variant type`.

Octachron commented on typing/printtyp.ml:
```diff
@@ -2176,51 +2218,96 @@ let unification_error env tr txt1 ppf txt2 ty_expect_explanation =
+    env tr trace_format.incompatibility_phrase mis txt1 ppf txt2 ty_expect_explanation
+
+let report_error trace_format ppf env tr
+      ?(type_expected_explanation = fun _ -> ())
```
The `type_expected_explanation`  argument is only used in `Typecore` in the `unification` case. Do we want to expose it for all trace formats?

antalsz replied:
I'm not sure how to avoid it – what did you have in mind?

Octachron replied:
A simple option would be to have a separate function for unification error report with an expected type explanation since the `report_error` function is mostly glue code. However, I am not sure if this is a good idea or not.

antalsz replied:
I decided to go with this change, or at least a variant of it – we now expose `report_unification_error`, `report_equality_error`, and `report_moregen_error` as separate functions, with the unification case having the optional argument and the latter two not having it.  As a side effect, I also no longer expose the `trace_format` type.

Octachron reviewed (commented):
I have a handful of remarks (mostly agreeing with @lpw25 's comments) on `Printtyp`, but otherwise the change in `Printtyp` looks good.

antalsz committed:
Address all the excellent reviewer suggestions

One test changes here, reverting to its pre-patch format, as a result
of some unification of different trace printing behavior; the prior
error message was considered acceptable as well, and both contain the
same amount of information, so this doesn't seem to be an issue.

antalsz commented:
All the comments above were very helpful!  I've addressed almost all of them now – sorry about the delay.  None of the comments that remain outstanding seem major; they all seem like easily-clarified and quickly-addressable points.  Let me know if there's anything else we're looking for!

antalsz committed:
Change `Printtyp.trace_format` into a GADT and expose it

antalsz committed:
Return `prepare_expansion_head` to the global scope

antalsz committed:
Replace `report_error` and `trace_format` with 3 separate functions

antalsz committed:
Fix minor bug in `Ctype.full_expand` and document invariant

antalsz committed:
Fix names of arguments to `Ctype.{,raise_}scope_escape_exn`

antalsz commented:
I have addressed the last couple of comments and rebased onto trunk.  I believe this PR is ready to be merged.

One note is that I fixed a small bug in `Ctype.full_expand`, which needed to return a physically equal value if no expansion happened; this showed up due to the interaction between the changes in this commit and the new changes that I rebased onto.

antalsz committed:
Add Changes entry for #10170

lpw25 reviewed (approved):
I think this is ready to merge

Octachron commented:
There is still some formatting errors: a missing header for the new `Errortrace` module, many overlong lines, and a non-ascii `…` in Printtyp. 

Speaking about copyright header, @antalsz did you sign the CLA?

Once those minor wrinkles are fixed, I would be happy to merge.

lpw25 commented:
@antalsz is covered as part of Jane Street for the CLA.

antalsz committed:
Add missing copyright headers to `errortrace.ml{,i}`

The code was pulled out of `ctype.ml{,i}`; per Florian, he wrote the
code for `Unification_trace` in 2018.

antalsz committed:
Fix formatting issues

* Fit to 80-column line width
* Remove stray non-ASCII character
* Delete stray empty line at end of file

antalsz commented:
Thanks for the formatting and copyright help, @Octachron!  I have updated the line widths, copied the copyright headers from `ctype.ml{,i}` (since that's where the `errortrace.ml{,i}` code came from), and deleted the non-ASCII character.  I then also rebased back onto `trunk` since there was a formatting conflict in a Makefile.  Since I'm covered by the Jane Street's CLA, I think that's everything!

Octachron commented on typing/errortrace.ml:
```diff
@@ -1,3 +1,18 @@
+(*                                                                        *)
+(*  Xavier Leroy and Jerome Vouillon, projet Cristal, INRIA Rocquencourt  *)
+(*                                                                        *)
+(*   Copyright 1996 Institut National de Recherche en Informatique et     *)
```
Looking at `git blame`, it seems that I wrote the corresponding code for the `Unification_trace` (in 2018). I would rather kept the copyright header not wildly inaccurate (the 1996 date in particular does seem strange). Could you also squash all check-typo commits into one?

antalsz replied:
This makes sense – I've made these changes, so let me know if you're satisfied that the new copyright headers are to your satisfaction!

Octachron replied:
That's look good (if the format is accepted by check-typo). You also need to rebase due to the change in toplevel/genprintval (sorry about that); then we can merge before another conflicting change sneak in.

lpw25 commented:
I'm afraid you lost the race with #10396, so you'll need to rebase again.

antalsz commented:
@lpw25 The rebase is done, thanks for the heads-up

gasche commented:
Thanks @antalsz for the nice work, and @lpw25 and @Octachron for the nice reviews! Merging.

gasche commented:
@Vanille-N just reported a regression caused by this PR in #12971. Repro case:

```ocaml
type 'a t = T of 'a

module Seq : sig
  type 'a t = unit -> 'a node
  and 'a node
  
  val empty : 'a t
  val cons : 'a -> 'a t -> 'a t
end = struct
  type 'a t = unit -> 'a node
  and 'a node = unit
  
  let empty () = ()
  let cons x xs () = ()
end 

let to_seq (xt : 'a t) : 'a Seq.t =
  let T x = xt in
  Seq.cons Seq.empty x
```

Behavior:

```
$ ocamlc -i test.ml
Fatal error: exception Ctype.Escape(_)
```

I can fix the issue with the following change (right on top of this PR):

```diff
diff --git i/typing/ctype.ml w/typing/ctype.ml
index da63dd803c..f2d81cc25a 100644
--- i/typing/ctype.ml
+++ w/typing/ctype.ml
@@ -1792,7 +1792,7 @@ let rec occur_rec env allow_recursive visited ty0 = function
         let visited = TypeSet.add ty visited in
         iter_type_expr (occur_rec env allow_recursive visited ty0) ty
       with Occur -> try
-        let ty' = try_expand_head try_expand_once env ty in
+        let ty' = try_expand_head try_expand_safe env ty in
         (* This call used to be inlined, but there seems no reason for it.
            Message was referring to change in rev. 1.58 of the CVS repo. *)
         occur_rec env allow_recursive visited ty0 ty'
```

Could one of you (for example @antalsz or @Octachron) correct that this is a correct fix for the issue? 

I see another call to `try_expand_once_opt` left over in typedecl.ml (`check_well_founded`); I wondered if that one was also wrong, but in fact it looks like it may be intentional as `check_well_founded` is catching `Escape _` .

gasche commented:
I spoke slightly too soon: the issue cannot be reproduced anymore in 5.2 or trunk, due to the following change between 5.1 and 5.2 which makes it go away: 7d6bd421a9d7af9fd5bd482736cf4af933598b69 .

I wonder if we still do want to fix the leftover `try_expand_once` anyway.
